### ISSUE TITLE
Explicit `T: Archive` bounds

### ISSUE THREAD

**Issue by bonsairobo** [bonsairobo/rkyv_impl#1]
Right now, the user must provide `T: Archive` bounds for generics, but this unnecessarily constrains the non-archived `impl`, for example:

```rust
#[archive_impl]
impl<T> Foo<T>
where
    T: Archive<Archived = T>,
{
    fn get_slice(&self) -> &[T] {
        &self.field
    }
}
```

This leaves the bound on `T` within the `impl<T> Foo<T> ... { ... }` item, but the bound is really only required for the `impl<T> ArchivedFoo<T>` item.

We should find a way to limit the placement of these bounds, either with another macro annotations or by placing them automatically. But automatic placement may go too far in the opposite direction; we don't always need `T: Archive` bounds, for example:

```rust
struct Foo<T> {
    marker: PhantomData<T>
}
```

**Comment by bonsairobo** [comment]
To characterize the problem more completely, we need to consider not only the bounds to be applied, but also the actual types being used within the `impl` block.

By the nature of the `Archive` derive, the types of all fields on `Self` (and any recursively contained types) have the possibility of being `Archived<FieldT>`.

If `FieldT` is ever referenced (e.g. via type annotation or return type) in a place where `Archived<FieldT> != FieldT`, then we are in a pretty difficult situation from the macro's perspective. Determining whether `FieldT` needs to be replaced with `Archived<FieldT>` seems infeasible in general, as we would need to parse the expression referring to `FieldT` and determine that it is in fact referring to a field on `self` and not just some other variable. I think this situation will remain incompatible with the macro in the foreseeable future.

Excluding the above scenario, we only require that the field types and their archived counterparts have compatible API usage within the `impl`. This might involve requiring a trait bound `T: Archive<Archived = T>`, so we should support specifying this bound on the generated `impl` block.

**Comment by bonsairobo** [comment]
Another question: once we have something like this:

```rust
#[archive_impl(bound("T: Archive"))]
impl<T> Foo<T> where T: PartialEq { ... }
```

Should the `where` clause be copied onto the generated `impl`? This would make sense when `T: Archive<Archived = T>`, but I'm less sure about other scenarios.

I think for now it will be convenient to copy the bounds, but eventually we may encounter a reason to omit them.

**Comment by bonsairobo** [comment]
For now, this is fixed by `d963dba`
